jib-core/src/main/java/com/google/cloud/tools/jib/configuration/LayerConfiguration.java

```diff
@@ -128,13 +137,67 @@ public Builder addEntry(
      */
     public Builder addEntryRecursive(Path sourceFile, AbsoluteUnixPath pathInContainer)
         throws IOException {
+      return addEntryRecursive(
+          sourceFile, pathInContainer, LayerEntry.DEFAULT_FILE_PERMISSIONS_PROVIDER);
+    }
+
+    /**
+     * Adds an entry to the layer. If the source file is a directory, the directory and its contents
+     * will be added recursively.
+     *
+     * @param sourceFile the source file to add to the layer recursively
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @param filePermissionProvider a provider that takes a source file and returns the file
+     *     permissions that should be set when it is added to the container
+     * @return this
+     * @throws IOException if an exception occurred when recursively listing the directory
+     */
+    public Builder addEntryRecursive(
+        Path sourceFile,
+        AbsoluteUnixPath pathInContainer,
+        Function<Path, FilePermissions> filePermissionProvider)
+        throws IOException {
+      return addEntryRecursive(
+          sourceFile,
+          pathInContainer,
+          filePermissionProvider,
+          LayerEntry.DEFAULT_FILE_TIMESTAMP_PROVIDER);
+    }
+
+    /**
+     * Adds an entry to the layer. If the source file is a directory, the directory and its contents
+     * will be added recursively.
+     *
+     * @param sourceFile the source file to add to the layer recursively
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @param filePermissionProvider a provider that takes a source file and returns the file
+     *     permissions that should be set when it is added to the container
+     * @param fileTimestampProvider a provider that takes a source file and returns the file
```

Comment:
I think these functions should act on pathInContainer, not sourceFile. For example, you want to put a lot files in a source directory recursively and you want to set an executable bit only on a few script files. Our plugin configurations take the same approach. With the per-sourceFile way, you end up having to set an executable bit on every file under the directory.

Reply:
Decided to take both source and destination, since both may be useful (e.g. our default permissions provider needs to check if the source is a file or folder to give 644 and 755 respectively).